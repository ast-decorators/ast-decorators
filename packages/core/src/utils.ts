import {ASTDecoratorTransformerOptions} from '@ast-decorators/typings';
import {BabelFileResult, NodePath} from '@babel/core';
import {Binding} from '@babel/traverse';
import {
  CallExpression,
  Decorator,
  Identifier,
  ImportDeclaration,
  ImportDefaultSpecifier,
  ImportNamespaceSpecifier,
  ImportSpecifier,
  isCallExpression,
  isMemberExpression,
  MemberExpression,
  StringLiteral,
} from '@babel/types';

export type ASTDecoratorExclusionOptions = {
  paths?: string[];
  names?: RegExp | string[];
};

export type ASTDecoratorCoreOptions = Readonly<{
  exclude?: ASTDecoratorExclusionOptions;
  transformers?: ASTDecoratorTransformerOptions;
}>;

export type PluginPass<T> = Readonly<{
  cwd: string;
  file: BabelFileResult;
  filename?: string;
  key: string;
  opts?: T;
}>;

const $args = Symbol('args');
const $binding = Symbol('binding');
const $decorator = Symbol('decorators');
const $id = Symbol('id');
const $isCall = Symbol('isCall');
const $object = Symbol('object');

const $handleDecoratorWithArgs = Symbol('handleDecoratorWithArgs');
const $handleDecoratorWithoutArgs = Symbol('handleDecoratorWithoutArgs');
const $handleMemberDecorator = Symbol('handleMemberDecorator');

export default class DecoratorMetadata {
  private [$args]?: readonly NodePath[];
  private [$binding]?: Binding;
  private [$decorator]: NodePath<Decorator>;
  private [$id]: NodePath<Identifier>;
  private [$isCall]: boolean;
  private [$object]?: NodePath<Identifier>;

  public constructor(decorator: NodePath<Decorator>) {
    this[$decorator] = decorator;
    const expression = decorator.get('expression');

    this[$isCall] = isCallExpression(expression);

    if (this[$isCall]) {
      this[$handleDecoratorWithArgs](expression as NodePath<CallExpression>);
    } else {
      this[$handleDecoratorWithoutArgs](
        expression as NodePath<MemberExpression | Identifier>,
      );
    }

    if (this.isFree) {
      this[$binding] = decorator.scope.getBinding(this[$id]!.node.name);
    } else {
      this[$binding] = decorator.scope.getBinding(this[$object]!.node.name);
    }
  }

  public get args(): readonly NodePath[] {
    return this[$args] ?? [];
  }

  public get binding(): Binding | undefined {
    return this[$binding];
  }

  public get bindingId(): NodePath<Identifier> {
    return this.isFree ? this[$id] : this[$object]!;
  }

  public get importSpecifier():
    | NodePath<
        ImportSpecifier | ImportNamespaceSpecifier | ImportDefaultSpecifier
      >
    | undefined {
    return this[$binding] ? (this[$binding]!.path as any) : undefined;
  }

  public get importSource(): NodePath<StringLiteral> | undefined {
    const {importSpecifier} = this;

    if (!importSpecifier) {
      return undefined;
    }

    const declaration = importSpecifier.parentPath as NodePath<
      ImportDeclaration
    >;

    return declaration.get('source');
  }

  public get isCall(): boolean {
    return this[$isCall];
  }

  /**
   * Is decorator identifier a part of MemberExpression?
   */
  public get isFree(): boolean {
    return !this[$object];
  }

  public get property(): string | undefined {
    return this.isFree ? undefined : this[$id].node.name;
  }

  public removeDecorator(): void {
    this[$decorator].remove();
  }

  public removeBinding(): void {
    if (!this[$binding]) {
      return;
    }

    this[$binding]!.referencePaths = this[$binding]!.referencePaths.filter(
      p => p !== this.bindingId,
    );

    if (this[$binding]!.referencePaths.length === 0) {
      const declaration = this.importSpecifier!.parentPath as NodePath<
        ImportDeclaration
      >;

      this.importSpecifier!.remove();

      if (declaration.node.specifiers.length === 0) {
        declaration.remove();
      }
    }
  }

  private [$handleDecoratorWithArgs](
    expression: NodePath<CallExpression>,
  ): void {
    this[$args] = expression.get('arguments');

    const callee = expression.get('callee');

    if (isMemberExpression(callee)) {
      this[$handleMemberDecorator](callee as NodePath<MemberExpression>);
    } else {
      this[$id] = callee as NodePath<Identifier>;
    }
  }

  private [$handleDecoratorWithoutArgs](
    expression: NodePath<MemberExpression | Identifier>,
  ): void {
    if (isMemberExpression(expression)) {
      this[$handleMemberDecorator](expression as NodePath<MemberExpression>);
    } else {
      this[$id] = expression as NodePath<Identifier>;
    }
  }

  private [$handleMemberDecorator](
    expression: NodePath<MemberExpression>,
  ): void {
    this[$object] = expression.get('object') as NodePath<Identifier>;
    this[$id] = expression.get('property') as NodePath<Identifier>;
  }
}
