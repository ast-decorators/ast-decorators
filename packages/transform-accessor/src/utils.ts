import {
  ASTClassMemberDecorator,
  ClassMemberMethod,
  ClassMemberProperty,
  PrivacyType,
} from '@ast-decorators/typings';
import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import {DecoratorSuitabilityFactors} from '@ast-decorators/utils/lib/checkDecoratorSuitability';
import createPropertyByPrivacy from '@ast-decorators/utils/lib/createPropertyByPrivacy';
import getMemberName from '@ast-decorators/utils/lib/getMemberName';
import {NodePath, template} from '@babel/core';
import {
  ArrowFunctionExpression,
  callExpression,
  CallExpression,
  Class,
  Decorator,
  functionDeclaration,
  FunctionExpression,
  identifier,
  Identifier,
  isArrowFunctionExpression,
  isClassPrivateProperty,
  isClassProperty,
  isFunctionExpression,
  isIdentifier,
  memberExpression,
  MemberExpression,
  PrivateName,
  thisExpression,
} from '@babel/types';

export type TransformAccessorOptions = Readonly<{
  transformerPath?: string;
  singleAccessorDecorators?: DecoratorSuitabilityFactors;
  privacy?: PrivacyType;
}>;

export type AccessorInterceptor = (value: any) => any;
export type AccessorInterceptorNode =
  | FunctionExpression
  | ArrowFunctionExpression
  | Identifier;

export type AccessorMethodCreatorOptions = Readonly<{
  allowThisContext: boolean;
  preservingDecorators: Decorator[] | null;
}>;

export type AccessorMethodCreator = (
  klass: NodePath<Class>,
  member: NodePath<ClassMemberProperty>,
  accessor: NodePath<AccessorInterceptorNode> | undefined,
  storage: Identifier | PrivateName,
  options: AccessorMethodCreatorOptions,
) => ClassMemberMethod;

export const assert = (
  decorator: string,
  member: NodePath<ClassMemberProperty>,
  accessors: ReadonlyArray<NodePath<AccessorInterceptorNode> | undefined>,
): void => {
  if (!isClassProperty(member) && !isClassPrivateProperty(member)) {
    throw new ASTDecoratorsError(
      `Applying @${decorator} decorator to something other than property is not allowed`,
    );
  }

  for (const accessor of accessors) {
    if (
      accessor &&
      !isFunctionExpression(accessor) &&
      !isArrowFunctionExpression(accessor) &&
      !isIdentifier(accessor)
    ) {
      throw new ASTDecoratorsError(
        'Accessor interceptor can only be function or a variable identifier',
      );
    }
  }
};

export const createStorage = (
  klass: NodePath<Class>,
  member: NodePath<ClassMemberProperty>,
  privacy: PrivacyType = 'hard',
): ClassMemberProperty =>
  createPropertyByPrivacy(
    privacy,
    String(getMemberName(member.node)),
    member.node.value,
    klass,
  );

const declarator = template(`const VAR = FUNCTION`);

export const generateInterceptor = (
  klass: NodePath<Class>,
  interceptor: AccessorInterceptorNode,
  type: 'get' | 'set',
): Identifier => {
  const isArrow = isArrowFunctionExpression(interceptor);
  const isRegular = isFunctionExpression(interceptor);

  const accessorId =
    isArrow || isRegular
      ? klass.parentPath.scope.generateUidIdentifier(`${type}Interceptor`)
      : (interceptor as Identifier);

  if (isArrow) {
    klass.insertBefore(
      declarator({
        FUNCTION: interceptor,
        VAR: accessorId,
      }),
    );
  } else if (isRegular) {
    klass.insertBefore(
      functionDeclaration(
        accessorId,
        (interceptor as FunctionExpression).params,
        (interceptor as FunctionExpression).body,
      ),
    );
  }

  return accessorId;
};

export const createAccessorDecorator = (
  decorator: string,
  accessor: NodePath<AccessorInterceptorNode> | undefined,
  impl: AccessorMethodCreator,
): ASTClassMemberDecorator<TransformAccessorOptions> => (
  klass: NodePath<Class>,
  member: NodePath<ClassMemberProperty>,
  {privacy}: TransformAccessorOptions = {},
): void => {
  assert(decorator, member, [accessor]);

  const storage = createStorage(klass, member, privacy);

  const method = impl(
    klass,
    member,
    accessor,
    storage.key as Identifier | PrivateName,
    {
      // TODO: Add option to set up context
      allowThisContext: true,
      preservingDecorators: member.node.decorators,
    },
  );

  member.replaceWithMultiple([storage, method]);
};

export const injectInterceptor = (
  klass: NodePath<Class>,
  interceptor: AccessorInterceptorNode,
  value: MemberExpression | Identifier,
  type: 'get' | 'set',
  allowThisContext: boolean,
): CallExpression => {
  const interceptorId = generateInterceptor(klass, interceptor, type);

  return isArrowFunctionExpression(interceptor) ||
    (isIdentifier(interceptor) && !allowThisContext)
    ? callExpression(interceptorId, [value])
    : callExpression(memberExpression(interceptorId, identifier('call')), [
        thisExpression(),
        value,
      ]);
};
