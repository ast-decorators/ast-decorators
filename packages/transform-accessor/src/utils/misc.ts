import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import {SuitabilityFactors} from '@ast-decorators/utils/lib/checkSuitability';
import {
  ASTClassMemberDecorator,
  ClassMemberMethod,
  ClassMemberProperty,
  PluginPass,
  PrivacyType,
} from '@ast-decorators/utils/lib/common';
import createPropertyByPrivacy from '@ast-decorators/utils/lib/createPropertyByPrivacy';
import getMemberName from '@ast-decorators/utils/lib/getMemberName';
import {NodePath, template} from '@babel/core';
import {
  ArrowFunctionExpression,
  callExpression,
  CallExpression,
  Class,
  cloneNode,
  Decorator,
  functionDeclaration,
  FunctionExpression,
  identifier,
  Identifier,
  isArrowFunctionExpression,
  isClassPrivateProperty,
  isClassProperty,
  isFunctionExpression,
  isIdentifier,
  isMemberExpression,
  memberExpression,
  MemberExpression,
  PrivateName,
  ThisExpression,
  thisExpression,
} from '@babel/types';
import shouldUseContext from './context';

export type TransformAccessorOptions = Readonly<{
  interceptorContext?: InterceptorContext;
  privacy?: PrivacyType;
  singleAccessorDecorators?: SuitabilityFactors;
  transformerPath?: string;
  useClassNameForStatic?: boolean;
}>;

export type AccessorInterceptor = (value: any) => any;
export type AccessorInterceptorNode =
  | FunctionExpression
  | ArrowFunctionExpression
  | Identifier
  | MemberExpression;

export type InterceptorContext = Readonly<{
  disableByDefault?: boolean;
  exclude?: SuitabilityFactors;
}>;

export type AccessorMethodCreatorOptions = Readonly<{
  useClassName: boolean;
  useContext: boolean;
  preservingDecorators: Decorator[] | null;
}>;

export type AccessorMethodCreator = (
  klass: NodePath<Class>,
  member: NodePath<ClassMemberProperty>,
  accessor: NodePath<AccessorInterceptorNode> | undefined,
  storage: Identifier | PrivateName,
  options: AccessorMethodCreatorOptions,
) => ClassMemberMethod;

export const assert = (
  decorator: string,
  member: NodePath<ClassMemberProperty>,
  interceptors: ReadonlyArray<NodePath<AccessorInterceptorNode> | undefined>,
): void => {
  if (!isClassProperty(member) && !isClassPrivateProperty(member)) {
    throw new ASTDecoratorsError(
      `Applying @${decorator} decorator to something other than property is not allowed`,
    );
  }

  for (const interceptor of interceptors) {
    if (
      interceptor &&
      !isFunctionExpression(interceptor) &&
      !isArrowFunctionExpression(interceptor) &&
      !isIdentifier(interceptor) &&
      !isMemberExpression(interceptor)
    ) {
      throw new ASTDecoratorsError(
        'Accessor interceptor can only be function, free variable or object property',
      );
    }
  }
};

export const createStorage = (
  klass: NodePath<Class>,
  member: NodePath<ClassMemberProperty>,
  privacy: PrivacyType = 'hard',
): ClassMemberProperty =>
  createPropertyByPrivacy(privacy, String(getMemberName(member.node)), klass, {
    // @ts-ignore
    static: member.node.static,
    value: member.node.value,
  });

const declarator = template(`const VAR = FUNCTION`);

export const generateInterceptor = (
  klass: NodePath<Class>,
  interceptor: AccessorInterceptorNode,
  type: 'get' | 'set',
): Identifier => {
  const isArrow = isArrowFunctionExpression(interceptor);
  const isRegular = isFunctionExpression(interceptor);

  const accessorId =
    isArrow || isRegular
      ? klass.parentPath.scope.generateUidIdentifier(`${type}Interceptor`)
      : (interceptor as Identifier);

  if (isArrow) {
    klass.insertBefore(
      declarator({
        FUNCTION: interceptor,
        VAR: accessorId,
      }),
    );
  } else if (isRegular) {
    klass.insertBefore(
      functionDeclaration(
        accessorId,
        (interceptor as FunctionExpression).params,
        (interceptor as FunctionExpression).body,
      ),
    );
  }

  return accessorId;
};

export const createAccessorDecorator = (
  decorator: string,
  interceptor: NodePath<AccessorInterceptorNode> | undefined,
  impl: AccessorMethodCreator,
): ASTClassMemberDecorator<TransformAccessorOptions> => (
  klass: NodePath<Class>,
  member: NodePath<ClassMemberProperty>,
  {
    interceptorContext,
    privacy,
    useClassNameForStatic,
  }: TransformAccessorOptions = {},
  {filename}: PluginPass,
): void => {
  assert(decorator, member, [interceptor]);

  const storage = createStorage(klass, member, privacy);

  const method = impl(
    klass,
    member,
    interceptor,
    storage.key as Identifier | PrivateName,
    {
      preservingDecorators: member.node.decorators,
      // @ts-ignore
      useClassName: !!member.node.static && !!useClassNameForStatic,
      useContext: shouldUseContext(interceptor, interceptorContext, filename),
    },
  );

  member.replaceWithMultiple([storage, method]);
};

export const ownerNode = (
  klass: NodePath<Class>,
  useClassName: boolean,
): Identifier | ThisExpression =>
  useClassName && klass.node.id ? cloneNode(klass.node.id) : thisExpression();

export const injectInterceptor = (
  klass: NodePath<Class>,
  interceptor: AccessorInterceptorNode,
  value: MemberExpression | Identifier,
  type: 'get' | 'set',
  useContext: boolean,
  useClassName: boolean,
): CallExpression => {
  const interceptorId = generateInterceptor(klass, interceptor, type);

  return isArrowFunctionExpression(interceptor) ||
    (isIdentifier(interceptor) && !useContext)
    ? callExpression(interceptorId, [value])
    : callExpression(memberExpression(interceptorId, identifier('call')), [
        ownerNode(klass, useClassName),
        value,
      ]);
};
