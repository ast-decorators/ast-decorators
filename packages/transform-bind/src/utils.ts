import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import {ClassMember} from '@ast-decorators/utils/lib/common';
import {isClassMethod, isClassPrivateMethod} from '@babel/types';

export type TransformBindOptions = Readonly<{
  transformerPath?: string;
}>;

export const assert = (decorator: string, member: ClassMember): void => {
  if (!isClassMethod(member) && !isClassPrivateMethod(member)) {
    throw new ASTDecoratorsError(
      `Applying @${decorator} decorator to something other than method is not allowed`,
    );
  }
};
