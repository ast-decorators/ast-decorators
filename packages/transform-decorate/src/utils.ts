export type TransformBindOptions = Readonly<{
  transformerPath?: string;
}>;
