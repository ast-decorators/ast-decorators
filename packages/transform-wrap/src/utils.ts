import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import getMemberName from '@ast-decorators/utils/lib/getMemberName';
import type {ClassMemberMethod} from '@ast-decorators/utils/src/common';
import {
  ArrowFunctionExpression,
  FunctionExpression,
  Identifier,
  isClassPrivateProperty,
  isClassProperty,
  isLiteral,
  MemberExpression,
} from '@babel/types';

export type TransformWrapOptions = Readonly<{
  transformerPath?: string;
}>;

export type AllowedWrappers =
  | FunctionExpression
  | ArrowFunctionExpression
  | Identifier
  | MemberExpression;

export const assertWrap = (
  wrapperFunction: AllowedWrappers | undefined,
  member: ClassMemberMethod,
): void => {
  if (!wrapperFunction) {
    const methodName = getMemberName(member);

    throw new ASTDecoratorsError(
      `@wrap: No wrapper function provided${
        methodName !== undefined ? ` for ${methodName}` : ''
      }`,
    );
  }

  if (
    (isClassProperty(member) || isClassPrivateProperty(member)) &&
    (!member.value || isLiteral(member.value))
  ) {
    throw new ASTDecoratorsError(
      `@wrap can only be applied to class methods or properties with ` +
        'functions assigned',
    );
  }
};
