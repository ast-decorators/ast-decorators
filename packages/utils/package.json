{
  "name": "@ast-decorators/utils",
  "version": "0.1.0",
  "description": "Common utils for AST decorators",
  "types": "lib/index.d.ts",
  "scripts": {
    "build": "npm run build:remove && run-p build:js build:dts",
    "build:js": "babel src -d lib -x .ts --ignore src/**/*.d.ts --copy-files --config-file ../../babel.config.json",
    "build:dts": "tsc -p tsconfig.build.json --emitDeclarationOnly",
    "build:remove": "rimraf lib",
    "typecheck": "tsc -p tsconfig.json --noEmit"
  },
  "dependencies": {
    "@ast-decorators/typings": "^0.1.0",
    "@babel/types": "^7"
  }
}
