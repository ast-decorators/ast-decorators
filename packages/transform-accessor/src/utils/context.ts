import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import checkSuitability from '@ast-decorators/utils/lib/checkSuitability';
import {ImportMetadata} from '@ast-decorators/utils/lib/metadata';
import {NodePath} from '@babel/core';
import {
  Identifier,
  isArrowFunctionExpression,
  isFunctionDeclaration,
  isFunctionExpression,
  isImportDeclaration,
  isMemberExpression,
  isVariableDeclarator,
  MemberExpression,
} from '@babel/types';
import {AccessorInterceptorNode, InterceptorContext} from './misc';

const shouldUseContext = (
  fn: NodePath<AccessorInterceptorNode> | undefined,
  {disableByDefault = false, exclude = {}}: InterceptorContext = {},
  filename: string,
): boolean => {
  if (!fn) {
    return false;
  }

  // If it is a declared regular function, context should be used.
  if (isFunctionExpression(fn)) {
    return true;
  }

  // Is it is a declared arrow function, context is unnecessary.
  if (isArrowFunctionExpression(fn)) {
    return false;
  }

  const isMember = isMemberExpression(fn);

  const id = isMember
    ? (fn.get('object') as NodePath<Identifier>)
    : (fn as NodePath<Identifier>);

  const binding = fn.scope.getBinding(id.node.name);

  if (!binding) {
    throw new ASTDecoratorsError(`${id.node.name} is not defined`);
  }

  let name: string | undefined;
  let source: string | undefined;

  const {path: declarationOrSpecifier} = binding;

  if (isImportDeclaration(declarationOrSpecifier.parentPath)) {
    // If the callback is imported, let's use global transformer rules.

    const {importSource, originalImportName} = new ImportMetadata(
      fn as NodePath<MemberExpression | Identifier>,
    );

    name = originalImportName;
    source = importSource?.node.value;
  } else if (!isMember) {
    // If element is declared in the same file and it is an arrow or a regular
    // function, let's decide if we need to add context.

    if (isVariableDeclarator(declarationOrSpecifier)) {
      const init = declarationOrSpecifier.get('init');

      if (isFunctionExpression(init)) {
        return true;
      } else if (isArrowFunctionExpression(init)) {
        return false;
      }
    } else if (isFunctionDeclaration(declarationOrSpecifier)) {
      return true;
    }

    throw new ASTDecoratorsError(`${id.node.name} is not a function`);
  } else {
    // If our function is a part of an object, let's use global transformer
    // settings.
    ({name} = (fn.get('property') as NodePath<Identifier>).node);
  }

  const suitable = checkSuitability({name, source}, exclude, filename);

  // If context is disabled by default, all functions that fit "exclude" option
  // should use context. Otherwise, exclusions should not use context.
  return disableByDefault ? suitable : !suitable;
};

export default shouldUseContext;
