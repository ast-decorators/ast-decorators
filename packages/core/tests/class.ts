import {compare as _compare} from '../../../utils/testing';

describe('@ast-decorators/core', () => {
  describe('class', () => {
    const compare = async (fixture: string): Promise<void> =>
      _compare(__dirname, 'class', fixture);

    it('compiles decorator for a class', async () => {
      await compare('default');
    });

    it('compiles multiple decorators for a class', async () => {
      await compare('multiple-decorators');
    });

    it('compiles decorators imported as namespace', async () => {
      await compare('namespace');
    });

    it('compiles renamed decorators', async () => {
      await compare('rename');
    });

    it('compiles decorators with params', async () => {
      await compare('params');
    });
  });
});
