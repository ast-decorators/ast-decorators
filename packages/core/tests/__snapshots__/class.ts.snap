// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@ast-decorators/core class compiles decorator for a class 1`] = `
"class Link extends HTMLAnchorElement {}

customElements.define('x-test', Link);"
`;

exports[`@ast-decorators/core class compiles decorators imported as namespace 1`] = `
"class Link extends HTMLAnchorElement {}

console.log('foo is 1');
console.log('bar is 0');"
`;

exports[`@ast-decorators/core class compiles decorators with params 1`] = `
"class Link extends HTMLAnchorElement {}

console.log([3, 4]);
console.log(2);
console.log(1);
console.log('test');
console.log({
  bar: 'baz'
});"
`;

exports[`@ast-decorators/core class compiles multiple decorators for a class 1`] = `
"class Link extends HTMLAnchorElement {}

console.log('foo is 1');
console.log('bar is 0');"
`;

exports[`@ast-decorators/core class compiles renamed decorators 1`] = `
"class Link extends HTMLAnchorElement {}

console.log('foo is 1');
console.log('bar is 0');"
`;

exports[`@ast-decorators/core class ignores other decorators 1`] = `
"import anotherDecorator from 'other-decorators';

@anotherDecorator
class Link extends HTMLAnchorElement {}

console.log('foo is 1');
console.log('bar is 0');"
`;
